extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,242 @@
+    url = urlReplacements.expandSync(
+        url, macros, undefined, whitelist);
+  }
+  if (!isSecureUrl(url)) {
+    return logAndAddErrorResponse_(promiseArray, RTC_ERROR_ENUM.INSECURE_URL,
+        opt_vendor || url);
+  }
+  if (seenUrls[url]) {
+    return logAndAddErrorResponse_(promiseArray, RTC_ERROR_ENUM.DUPLICATE_URL,
+        opt_vendor || url);
+  }
+  seenUrls[url] = true;
+  promiseArray.push(sendRtcCallout_(
+      url, rtcStartTime, win, timeoutMillis, opt_vendor || url));
+}
+
+/**
+ * @param {!string} url
+ * @param {!number} rtcStartTime
+ * @param {!Window} win
+ * @param {!number} timeoutMillis
+ * @param {!string} callout
+ * @return {!Promise<!rtcResponseDef>}
+ * @private
+ */
+function sendRtcCallout_(
+    url, rtcStartTime, win, timeoutMillis, callout) {
+  /**
+   * Note: Timeout is enforced by timerFor, not the value of
+   *   rtcTime. There are situations where rtcTime could thus
+   *   end up being greater than timeoutMillis.
+   */
+  return Services.timerFor(win).timeoutPromise(
+      timeoutMillis,
+      Services.xhrFor(win).fetchJson(
+          // NOTE(bradfrizzell): we could include ampCors:false allowing
+          // the request to be cached across sites but for now assume that
+          // is not a required feature.
+          url, {credentials: 'include'}).then(res => {
+            return res.text().then(text => {
+              const rtcTime = Date.now() - rtcStartTime;
+              // An empty text response is allowed, not an error.
+              if (!text) {
+                return {rtcTime, callout};
+              }
+              const rtcResponse = tryParseJson(text);
+              return rtcResponse ? {rtcResponse, rtcTime, callout} :
+              buildErrorResponse_(
+                  RTC_ERROR_ENUM.MALFORMED_JSON_RESPONSE, callout, rtcTime);
+            });
+          })).catch(unusedError => {
+            return buildErrorResponse_(RTC_ERROR_ENUM.NETWORK_FAILURE,
+                callout, Date.now() - rtcStartTime);
+          });
+}
+
+/**
+ * Attempts to parse the publisher-defined RTC config off the amp-ad
+ * element, then validates that the rtcConfig exists, and contains
+ * an entry for either vendor URLs, or publisher-defined URLs. If the
+ * config contains an entry for timeoutMillis, validates that it is a
+ * number, or converts to a number if number-like, otherwise overwrites
+ * with the default.
+ * IMPORTANT: If the rtcConfig is invalid, RTC is aborted, and the ad
+ *   request continues without RTC.
+ * @param {!Element} element
+ * @return {?Object}
+ * @visibleForTesting
+ */
+export function validateRtcConfig_(element) {
+  const defaultTimeoutMillis = 1000;
+  const rtcConfig = tryParseJson(
+      element.getAttribute('rtc-config'));
+  if (!rtcConfig) {
+    return null;
+  }
+  const rtcConfigKeyWhitelist = {
+    urls: true,
+    vendors: true,
+    timeoutMillis: true,
+  };
+  Object.keys(rtcConfig).forEach(key => {
```

Comment:
I still think a switch would be better here:

try {
   user().assert(rtcConfig['vendors'] || rtcConfig['urls'],
        'RTC Config must specify vendors or urls');
  Object.key(rtcConfig).forEach(key => {
     const val = rtcConfig[key];
     switch(key) {
        case 'urls':
           user().assert(isArray(value), '...');
           break;
        ...
        default:
           user().warn(TAG, `Unknownn RTC Config key: ${key}`); 
      }
  });
} catch (unused_err) {
   return false;
}
return true;
});

Reply:
done